### PR TITLE
Initial TCP Support

### PR DESCRIPTION
**Pull request by nibanks**
Adds TCP support to the data path abstraction layer. This is so that we can add TCP support to our performance tool to compare with QUIC. Only supported on Windows user mode. Because it no longer abstracts just a UDP binding, renames the data path binding to socket.

### CODE REVIEW

File: src/inc/quic_datapath.h [hunk 1]
```diff
@@ -306,63 +361,89 @@ QuicDataPathResolveAddress(
...
     _In_opt_ const QUIC_ADDR* LocalAddress,
     _In_opt_ const QUIC_ADDR* RemoteAddress,
-    _In_opt_ void* RecvCallbackContext,
-    _Out_ QUIC_DATAPATH_BINDING** Binding
+    _In_opt_ void* CallbackContext,
+    _Out_ QUIC_SOCKET** Socket
+    );
+
+//
+// Creates a TCP socket for the (optional) (optional) given local address and
```
Comment (ThadHouse) [inline src/inc/quic_datapath.h:382]: (optional) (optional) is a bit weird here. Also doesn't match the UDP docs.

Comment (nibanks) [reply src/inc/quic_datapath.h:382]: Hmm, I think a copy/paste bug. I'll fix it up.

File: src/core/library.c [hunk 2]
```diff
@@ -191,6 +191,10 @@ MsQuicLibraryInitialize(
     QUIC_STATUS Status = QUIC_STATUS_SUCCESS;
     BOOLEAN PlatformInitialized = FALSE;
     uint32_t DefaultMaxPartitionCount = QUIC_MAX_PARTITION_COUNT;
+    const QUIC_UDP_DATAPATH_CALLBACKS DatapathCallbacks = {
```
Comment (anrossi) [inline src/core/library.c:194]: Even if these are const, are they still stack allocated? shouldn't this be global?

Comment (nibanks) [reply src/core/library.c:194]: No need. The datapath copies the contents.

File: src/core/quicdef.h [hunk 3]
```diff
@@ -234,7 +234,7 @@ QUIC_STATIC_ASSERT(IS_POWER_OF_TWO(QUIC_MAX_RANGE_DECODE_ACKS), L"Must be power
 // Path MTU discovery will always start with/initialize with the smallest
 // allowable MTU for QUIC (1280 bytes).
 //
-#define QUIC_DEFAULT_PATH_MTU                   QUIC_MIN_MTU
+#define QUIC_DEFAULT_PATH_MTU                   1280    // TODO - Use 1200 instead
```
Comment (ThadHouse) [inline src/core/quicdef.h:237]: Is this TODO going to be fixed in this PR, or for later?

**Review by ThadHouse — APPROVED** [review]

File: src/platform/datapath_winuser.c [hunk 4]
```diff
@@ -573,11 +747,27 @@ QuicDataPathInitialize(
...
+    }
+    if (TcpCallbacks != NULL) {
+        if (TcpCallbacks->Accept == NULL ||
+            TcpCallbacks->Connect == NULL ||
+            TcpCallbacks->Receive == NULL) {
+            Status = QUIC_STATUS_INVALID_PARAMETER;
+            Datapath = NULL;
+            goto Exit;
+        }
+    }
```
Comment (anrossi) [inline src/platform/datapath_winuser.c:770]: enforce that specifying both Udp and Tcp callbacks should fail. A socket cannot be both TCP and UDP, right?

File: src/platform/datapath_winuser.c [hunk 5]
```diff
@@ -606,14 +796,21 @@ QuicDataPathInitialize(
     }
 
     RtlZeroMemory(Datapath, DatapathLength);
-    Datapath->RecvHandler = RecvCallback;
-    Datapath->UnreachableHandler = UnreachableCallback;
+    if (UdpCallbacks) {
+        Datapath->UdpHandlers = *UdpCallbacks;
```
Comment (anrossi) [inline src/platform/datapath_winuser.c:800]: This requires the upper layer (QUIC core) to keep this structure alive for the duration of the datapath layer, which kind of breaks the layering abstraction between them.  Shouldn't this copy the structure/pointers locally?

Comment (ThadHouse) [reply src/platform/datapath_winuser.c:800]: The deref will copy the struct, keeping a local copy of the pointers.

Comment (anrossi) [reply src/platform/datapath_winuser.c:800]: Ah, I (incorrectly) assumed it was a pointer-pointer. You're right.

**Review by anrossi — APPROVED** [review]

File: src/core/quicdef.h (hunk 3 above)
Comment (nibanks) [reply src/core/quicdef.h:237]: Later.

File: src/platform/datapath_winuser.c (hunk 4 above)
Comment (nibanks) [reply src/platform/datapath_winuser.c:770]: This is the datapath layer, not socket layer. You can have one datapath handle that you use to create both UDP and TCP sockets.

**Pull request merged**